addons/godot-xr-tools/staging/staging.gd

```diff
@@ -0,0 +1,237 @@
+var current_scene_path : String
+
+var _tween : Tween
+
+func _on_exit_to_main_menu():
+	load_scene(main_scene)
+
+func _on_load_scene(p_scene_path : String):
+	load_scene(p_scene_path)
+
+func _add_signals(p_scene : XRToolsSceneBase):
+	p_scene.connect("exit_to_main_menu", _on_exit_to_main_menu)
+	p_scene.connect("load_scene", _on_load_scene)
+
+func _remove_signals(p_scene : XRToolsSceneBase):
+	p_scene.disconnect("exit_to_main_menu", _on_exit_to_main_menu)
+	p_scene.disconnect("load_scene", _on_load_scene)
+
+func load_scene(p_scene_path : String):
+	# Do not load if in the editor
+	if Engine.is_editor_hint():
+		return
+
+	# Check if it's already loaded...
+	if p_scene_path == current_scene_path:
+		return
+
+	if !xr_origin:
+		return
+
+	if !xr_camera:
+		return
+
+	if current_scene:
+		# Start by unloading our scene
+
+		# Let the scene know we're about to remove it
+		current_scene.scene_pre_exiting()
+
+		# Remove signals
+		_remove_signals(current_scene)
+
+		# Fade to black
+		if _tween:
+			_tween.kill()
+		_tween = get_tree().create_tween()
+		_tween.tween_method(set_fade, 0.0, 1.0, 1.0)
+		await _tween.finished
+
+		# Now we remove our scene
+		emit_signal("scene_exiting", current_scene)
+		current_scene.scene_exiting()
+		$Scene.remove_child(current_scene)
+		current_scene.queue_free()
+		current_scene = null
+
+		# Make our loading screen visible again and reset some stuff
+		xr_origin.set_process_internal(true)
+		xr_camera.current = true
+		$WorldEnvironment.environment = loading_screen_environment
+		$LoadingScreen.progress = 0.0
+		$LoadingScreen.enable_press_to_continue = false
+		$LoadingScreen.follow_camera = true
+		$LoadingScreen.visible = true
+		emit_signal("switching_to_loading_scene")
+
+		# Fade to visible
+		if _tween:
+			_tween.kill()
+		_tween = get_tree().create_tween()
+		_tween.tween_method(set_fade, 1.0, 0.0, 1.0)
+		await _tween.finished
+
+	# Attempt to load our scene
+	#ResourceQueue.queue_resource(p_scene_path)
+	#while !ResourceQueue.is_ready(p_scene_path):
+	#	# wait one second
+	#	await get_tree().create_timer(1.0).timeout
+	#	
+	#	$LoadingScreen.progress = ResourceQueue.get_progress(p_scene_path)
+
+	#var new_scene : PackedScene = ResourceQueue.get_resource(p_scene_path)
+	var new_scene : PackedScene = ResourceLoader.load(p_scene_path)
+
+	# Wait for user to be ready
+	if prompt_for_continue:
+		$LoadingScreen.enable_press_to_continue = true
+		await $LoadingScreen.continue_pressed
+
+	# Fade to black
+	if _tween:
+		_tween.kill()
+	_tween = get_tree().create_tween()
+	_tween.tween_method(set_fade, 0.0, 1.0, 1.0)
+	await _tween.finished	
+
+	# Hide our loading screen
+	$LoadingScreen.follow_camera = false
+	$LoadingScreen.visible = false
+
+	# Turn off internal process on our FPController, the internal process
+	# of our XROrigin3D will submit its positioning data to the XRServer.
+	# With two XROrigin3D nodes we'll get competing data.
+	xr_origin.set_process_internal(false)
```

Comment:
So this is no longer needed, instead we need to make the Origin node loaded in current. We can put that bit of code in `scene_base.gd/scene_loaded()`

Reply:
That said, there is a bug in the current beta (already fixed in master) where we're not updating the starting position. In the new situation we're only informing the XRServer when the current origin node's position changes, instead of doing this every frame.